### PR TITLE
Introduce a function to compute the doubling/halving time for a lightcurve

### PR DESCRIPTION
**Pull request by cgalelli**
Similarly to pr #4501 and pr #4726, I wrote a script providing the computation of the doubling/halving time, useful for studies of temporal variability in lightcurves (example: AGN flares)

This PR includes the code for the computation of the variable itself in stats/variability.py, as well as a wrapper in estimators/utils.py, each with its test.

### CODE REVIEW

File: gammapy/stats/variability.py [hunk 1]
```diff
@@ -82,3 +84,76 @@ def compute_chisq(flux):
...
+        * np.log(2)
+        / flux[..., :-1]
+        * np.log(flux[..., 1:] / flux[..., :-1]) ** 2
+    )
+    times_err = np.sqrt(
+        (flux_err[..., 1:] * times_err_1) ** 2 + (flux_err[..., :-1] * times_err_2) ** 2
+    )
+
+    imin = np.argmin(
+        np.where(np.logical_and(np.isfinite(times), times > 0), times, 1e15 * u.s),
```
Comment (AtreyeeS) [inline gammapy/stats/variability.py:145]: you can use `np.infty` instead of `1e15`

Comment (AtreyeeS) [inline gammapy/stats/variability.py:145]: you can use `np.infty` instead of 1e15

File: gammapy/stats/variability.py [hunk 2]
```diff
@@ -82,3 +84,76 @@ def compute_chisq(flux):
...
+    ----------
+    ..[Brown2013] "Locating the γ-ray emission region
+    of the flat spectrum radio quasar PKS 1510−089", Brown et al. (2013)
+    https://academic.oup.com/mnras/article/431/1/824/1054498
+    """
+
+    flux = np.atleast_2d(flux).swapaxes(0, axis).T
+    flux_err = np.atleast_2d(flux_err).swapaxes(0, axis).T
+
+    times = (time[1:] - time[:-1]) / np.log2(flux[..., 1:] / flux[..., :-1])
```
Comment (AtreyeeS) [inline gammapy/stats/variability.py:127]: In this case, I suppose passing time edges is necessary, not time centers?

File: gammapy/stats/variability.py [hunk 3]
```diff
@@ -82,3 +84,76 @@ def compute_chisq(flux):
...
+        np.where(np.logical_and(np.isfinite(times), times < 0), times, -1e15 * u.s),
+        axis=-1,
+        keepdims=True,
+    )
+    index = np.concatenate([imin, imax], axis=-1)
+
+    dtime = np.take_along_axis(times, index, axis=-1)
+    dtime_err = np.take_along_axis(times_err, index, axis=-1)
+
+    return dtime, dtime_err
```
Comment (AtreyeeS) [inline gammapy/stats/variability.py:159]: Maybe it is also useful to know which flare/at what time this doubling scale is obtained. For a typical AGN light curve, you can have multiple flares...

File: gammapy/estimators/utils.py [hunk 4]
```diff
@@ -346,3 +347,55 @@ def compute_lightcurve_fvar(lightcurve, flux_quantity="flux"):
...
+        ],
+        names=(
+            "min_energy",
+            "max_energy",
+            "doubling_time",
+            "dtime_err",
+            "halving_time",
+            "htime_err",
+        ),
+        meta=lightcurve.meta,
```
Comment (AtreyeeS) [inline gammapy/estimators/utils.py:398]: I don't know if this meta info is relevant for the table...

File: gammapy/estimators/utils.py [hunk 5]
```diff
@@ -346,3 +347,55 @@ def compute_lightcurve_fvar(lightcurve, flux_quantity="flux"):
     )
 
     return table
+
+
+def compute_lightcurve_dtime(lightcurve, flux_quantity="flux"):
+    r"""Compute the doubling time of the input lightcurve.
```
Comment (AtreyeeS) [inline gammapy/estimators/utils.py:353]: Should mention that this is the minimum doubling time scale...

File: gammapy/stats/variability.py [hunk 6]
```diff
@@ -82,3 +84,76 @@ def compute_chisq(flux):
...
+    axis : int, optional
+        Axis along which the value is computed.
+        The default is to compute the value on axis 0.
+
+    Returns
+    -------
+    dtime, dtime_err : `~numpy.ndarray`
+        Characteristic doubling time, halving time and errors.
+
+    References
```
Comment (AtreyeeS) [inline gammapy/stats/variability.py:117]: Since users will be more exposed to the wrapper function, this detailed doc string can lie there..

**Review by AtreyeeS — COMMENTED** [review]
Thanks @cgalelli for this necessary functionality! 

One general comment is that `compute_dtime` might be more useful than simply only for light curves, eg, for flux profiles, it can be interesting to know the spatial scales at which the fluxes change. In this case, maybe

1. You can adapt `compute_dtime` to take any axis as an input. The naming can be adapted accordingly, eg, `compute_flux_doubling_scale` etc
2. The `compute_lightcure_dtime` can reside in `LightCurve` directly, rather than in utils. It improves the visibility...

**Comment by cgalelli** [comment]
> Thanks @cgalelli for this necessary functionality!
> 
> One general comment is that `compute_dtime` might be more useful than simply only for light curves, eg, for flux profiles, it can be interesting to know the spatial scales at which the fluxes change. In this case, maybe
> 
> 1. You can adapt `compute_dtime` to take any axis as an input. The naming can be adapted accordingly, eg, `compute_flux_doubling_scale` etc
> 2. The `compute_lightcure_dtime` can reside in `LightCurve` directly, rather than in utils. It improves the visibility...

Thanks for all the feedback @AtreyeeS ! I did not think about the expanded use cases for this function. I can surely make the necessary changes if you think this is interesting for more than just temporal variability. Do you think I should keep the "core" function in `variability.py`?

I'll also work on all the other points you raised. Thanks again!

File: gammapy/estimators/utils.py (hunk 4 above)
Comment (cgalelli) [reply gammapy/estimators/utils.py:398]: Absolutely, it is mostly a placeholder for future metadata or to not leave the table without

File: gammapy/stats/variability.py (hunk 1 above)
Comment (cgalelli) [reply gammapy/stats/variability.py]: Done, thanks!

File: gammapy/stats/variability.py (hunk 2 above)
Comment (AtreyeeS) [reply gammapy/stats/variability.py:127]: My point was that from `compute_lightcurve_dtime` you pass the time centers. As per your logic here, you should be passing the edges, no?

File: gammapy/stats/variability.py (hunk 3 above)
Comment (cgalelli) [reply gammapy/stats/variability.py]: Added in the new commit, thanks!

**Review by bkhelifi — APPROVED** [review]
Thanks @cgalelli . It seems good!

File: gammapy/stats/variability.py (hunk 2 above)
Comment (cgalelli) [reply gammapy/stats/variability.py]: Sorry for the delay in the response. I don't think any between edges (min or max) or centers are perfect. I would still prefer to pass the centers of the bins in this case as they better represent an "average coordinate" of the flux point. Sorry if I am missing something obvious, but why are edges more logic here?

**Review by bkhelifi — APPROVED** [review]
All good!

**Review by registerrier — CHANGES_REQUESTED** [review]
Thanks @cgalelli . This is a nice addition.

I think the implementation is too complex as is and should be simplified. Some suggestions:
- the idea of having the possibility to deal with other axes is interesting but I would keep that for the generic function working with numpy arrays/Quantity. The object working with `FluxPoints` should be kept simple and compute only flux doubling time scales.
- the `compute_flux_doubling` function implementation is too complex and should be clarified.
- there is no explicit expression given for the flux doubling time and its error in the reference given. If you have to derive it manually I think one should give the expression in the docstring.

As a more general comment, I wonder what the best approach could be. Technically, what is done is fitting a temporal model with one dof per time bin. Would it be possible or make sense to extract the information by fitting directly a dataset?

File: gammapy/estimators/utils.py [hunk 7]
```diff
@@ -346,3 +348,107 @@ def compute_lightcurve_fvar(lightcurve, flux_quantity="flux"):
     )
 
     return table
+
+
+def compute_lightcurve_flux_doubling(lightcurve, axis_name, flux_quantity="flux"):
```
[the remark this answers is not in the document]
Comment (cgalelli) [reply gammapy/estimators/utils.py]: Thanks Regis, in the latest commit I removed this possibility and left only the doubling time as a wrapper for FluxPoints objects. The base function can still use whatever axis.

File: gammapy/stats/variability.py [hunk 8]
```diff
@@ -82,3 +83,76 @@ def compute_chisq(flux):
...
+    axis : int, optional
+        Axis along which the value is computed.
+
+    Returns
+    -------
+    doubling, doubling_err, coord : `~numpy.ndarray`
+        Characteristic flux doubling, halving and errors,
+        with coordinates at which they were found.
+    """
+
```
[the remark this answers is not in the document]
Comment (cgalelli) [reply gammapy/stats/variability.py:120]: After researching possible alternatives, it seems in my opinion that axis swapping is still the best solution for this.

**Comment by registerrier** [comment]
Merging #4726 has introduced some merge conflicts unfortunately. 
Can you rebase @cgalelli ?

**Pull request closed without merging**